document/syntax/values.rst

```diff
@@ -0,0 +1,140 @@
+.. _syntax-value:
+.. index:: ! value
+   pair: abstract syntax; value
+
+Values
+------
+
+
+.. _syntax-byte:
+.. index:: ! byte
+   pair: abstract syntax; byte
+
+Bytes
+~~~~~
+
+The simplest form of value are raw uninterpreted *bytes*.
+In the abstract syntax they are represented as hexadecimal literals.
+
+.. math::
+   \begin{array}{llll}
+   \production{bytes} & \by &::=&
+     \hex{00} ~|~ \dots ~|~ \hex{FF} \\
+   \end{array}
+
+
+Conventions
+...........
+
+* The meta variable :math:`b` range over bytes.
+
+* The meta function :math:`\byte(n)` denotes the byte representing the natural number :math:`n < 256`.
+
+
+.. _syntax-int:
+.. _syntax-sint:
+.. _syntax-uint:
+.. index:: ! integer, ! unsigned integer, ! signed integer, ! uninterpreted integer
+   pair: abstract syntax; integer
+   pair: abstract syntax; unsigned integer
+   pair: abstract syntax; signed integer
+   pair: abstract syntax; uninterpreted integer
+   single: integer; unsigned
+   single: integer; signed
+   single: integer; uninterpreted
+
+Integers
+~~~~~~~~
+
+Different classes of *integers* with different value ranges are distinguished by their *size* and their *signedness*.
+
+.. math::
+   \begin{array}{llll}
+   \production{unsigned integers} & \uN &::=&
+     0 ~|~ 1 ~|~ \dots ~|~ 2^N{-}1 \\
+   \production{signed integers} & \sN &::=&
+     -2^{N-1} ~|~ \dots ~|~ {-}1 ~|~ 0 ~|~ 1 ~|~ \dots ~|~ 2^{N-1}{-}1 \\
+   \production{uninterpreted integers} & \iN &::=&
+     \uN ~|~ \sN \\
+   \end{array}
+
+The latter class defines *uninterpreted* integers, whose signedness interpretation can vary depending on context.
+A 2's complement conversion is assumed for values that are out-of-range for a chosen interpretation.
+That is, semantically, when interpreted as unsigned, negative values :math:`-n` convert to :math:`2^N-n`,
+and when interpreted as signed, positive values :math:`n \geq 2^{N-1}` convert to :math:`n-2^N`.
```

Comment:
Well, you don't want the AST to contain programs that are not actually expressible in the concrete syntax. Otherwise you gonna run into annoying impedance mismatches. For example, you'd have to specify what happens for out-of-bounds values (e.g. an offset >= 2^32) in places, or impose range restrictions on a meta level, which is more difficult to handle.

Signedness is especially relevant for semantics, think natural vs integer domains. In particular, most numbers in the AST are natural, and defining that directly avoids the need to say anything about negative values all over the place.

You may be right, though, that expressing this in terms of signedness is mixing levels a little. It might make sense to distinguish static (all natural) numbers as they occur in the AST, from dynamic values like used for computation and only occurring in the AST with const instructions. Would naming the former "nat" make that distinction clearer?

The bit about conversion of uninterpreted integers is intended more as an explanation than a definition here. The actual definition will occur in the places of the spec where such values are actually produced/consumed. But I thought it's useful to mention upfront how they will be dealt with. But if the explanation is unclear it probably needs improvement. I reworded it somewhat, not sure it's better now. Also, it should link to the part of the spec that makes this more precise; I added a todo for that.